src/dune_rules/odoc_new.ml

```diff
@@ -0,0 +1,1960 @@
+    | [] -> "toplevel"
+    | xs -> String.concat ~sep:"." (List.rev_map xs ~f:ty_name)
+  ;;
+
+  let rec might_be_fallback = function
+    | [] -> false
+    | Private_lib _ :: _ -> false
+    | Package (Fallback, _) :: _ -> true
+    | Package (Unknown, _) :: _ -> true
+    | Package (Dune_with_modules, _) :: _ -> false
+    | Package (Local, _) :: _ -> false
+    | Subdir _ :: xs -> might_be_fallback xs
+  ;;
+
+  (* Used to suppress warnings on packages from the switch *)
+  let rec is_external = function
+    | [] -> false
+    | Private_lib _ :: _ -> false
+    | Package (Local, _) :: _ -> false
+    | Package (_, _) :: _ -> true
+    | Subdir _ :: xs -> is_external xs
+  ;;
+
+  let external_ty_to_dyn x =
+    let open Dyn in
+    match x with
+    | Unknown -> variant "Unknown" []
+    | Dune_with_modules -> variant "Dune_with_modules" []
+    | Fallback -> variant "Fallback" []
+    | Local -> variant "Local" []
+  ;;
+
+  let ty_to_dyn x =
+    let open Dyn in
+    match x with
+    | Private_lib lnu -> variant "Private_lib" [ String lnu ]
+    | Package (e, str) -> variant "Package" [ external_ty_to_dyn e; String str ]
+    | Subdir str -> variant "LocalSubLib" [ String str ]
+  ;;
+
+  let to_dyn x = Dyn.list ty_to_dyn x
+  let compare_ty x y = Dyn.compare (ty_to_dyn x) (ty_to_dyn y)
+
+  let subdir = function
+    | Package (_, str) -> str
+    | Private_lib lnu -> lnu
+    | Subdir str -> str
+  ;;
+
+  let local_dir : t -> string = fun v -> List.map ~f:subdir v |> String.concat ~sep:"/"
+
+  (* Where we find the odoc files for the indexes *)
+  let obj_dir ctx ~all : t -> Path.Build.t =
+    let root = Paths.root ctx ~all ++ "index" in
+    List.fold_right ~f:(fun x acc -> acc ++ subdir x) ~init:root
+  ;;
+
+  (* Where we find the output HTML files for artifacts that are children of
+     this index *)
+  let html_dir ctx ~all (m : t) =
+    let init = Paths.html_root ctx ~all ++ "docs" in
+    List.fold_right ~f:(fun x acc -> acc ++ subdir x) ~init m
+  ;;
+
+  (* Where we find odoc files for artifacts that are children of this index. *)
+  let odoc_dir ctx ~all (m : t) =
+    let init = Paths.root ctx ~all ++ "odoc" in
+    List.fold_right ~f:(fun x acc -> acc ++ subdir x) ~init m
+  ;;
+
+  let mld_name_ty : ty -> string = subdir
+
+  let mld_name : t -> string = function
+    | [] -> "docs"
+    | x :: _ -> mld_name_ty x
+  ;;
+
+  let mld_filename index = mld_name index ^ ".mld"
+  let mld_path ctx ~all index = obj_dir ctx ~all index ++ mld_filename index
+
+  (* Here we have the three constructors of the indexes. The first
+     takes an external lib found in a findlib directory: *)
+  let of_external_lib findlib_paths ty lib =
+    let local =
+      let local_full =
+        Lib.info lib
+        |> Lib_info.obj_dir
+        |> Obj_dir.dir
+        |> Paths.local_path_of_findlib_path findlib_paths
+      in
+      String.split ~on:'/' local_full
+    in
+    let init = [ Package (ty, List.hd local) ] in
```

Comment:
This won't work in general

Reply:
But see my other comments as I am dubious to the utility of the "local_path_of_findlib_path" function.